torch/csrc/autograd/variable.h

```diff
@@ -1,89 +1,99 @@
 #pragma once
 
+#include <mutex>
 #include <memory>
 #include <functional>
 #include <THPP/THPP.h>
 
 #include "torch/csrc/autograd/function.h"
-#include "torch/csrc/autograd/saved_variable.h"
+#include "torch/csrc/autograd/variable_version.h"
 #include "torch/csrc/Types.h"
 
 namespace torch { namespace autograd {
 
-struct VariableVersion;
+struct Variable : std::enable_shared_from_this<Variable> {
+
+  struct SavedVariable {
```

Comment:
Because `Variable` and `SavedVariable` depend on the structure of each other being defined (`Variable&` in the constructor of `SavedVariable`, `SavedVariable` returned from `Variable::save()`)